java-frontend/src/main/java/org/sonar/java/se/checks/DivisionByZeroCheck.java

```diff
@@ -0,0 +1,253 @@
+    super.scanFile(context);
+  }
+
+  private static class ZeroConstraint extends ObjectConstraint {
+    private ZeroConstraint(Tree syntaxNode, Status status) {
+      super(false, false, syntaxNode, status);
+    }
+
+    public static ZeroConstraint zero(Tree syntaxNode) {
+      return new ZeroConstraint(syntaxNode, Status.ZERO);
+    }
+
+    public static ZeroConstraint nonZero(Tree syntaxNode) {
+      return new ZeroConstraint(syntaxNode, Status.NON_ZERO);
+    }
+
+    @Override
+    public ZeroConstraint inverse() {
+      Tree syntaxNode = syntaxNode();
+      return isZero() ? nonZero(syntaxNode) : zero(syntaxNode);
+    }
+
+    private boolean isZero() {
+      return hasStatus(Status.ZERO);
+    }
+
+    @Override
+    public boolean isInversible() {
+      return isZero();
+    }
+
+    @Override
+    public String toString() {
+      // For debugging purpose
+      return "" + (isZero() ? Status.ZERO : Status.NON_ZERO);
+    }
+  }
+
+  @Override
+  public ProgramState checkPreStatement(CheckerContext context, Tree syntaxNode) {
+    if (syntaxNode.is(Tree.Kind.DIVIDE_ASSIGNMENT, Tree.Kind.DIVIDE, Tree.Kind.REMAINDER_ASSIGNMENT, Tree.Kind.REMAINDER)) {
+      ExpressionTree denominator;
+      SymbolicValue denominatorSV;
+      ProgramState programState = context.getState();
+      List<SymbolicValue> symbolicValues = programState.peekValues(2);
+      if (syntaxNode.is(Tree.Kind.REMAINDER_ASSIGNMENT, Tree.Kind.DIVIDE_ASSIGNMENT)) {
+        denominator = ((AssignmentExpressionTree) syntaxNode).expression();
+        denominatorSV = symbolicValues.get(1);
+      } else {
+        denominator = ((BinaryExpressionTree) syntaxNode).rightOperand();
+        denominatorSV = symbolicValues.get(0);
+      }
+
+      if (isZero(programState, denominatorSV)) {
+        String operation = syntaxNode.is(Tree.Kind.REMAINDER, Tree.Kind.REMAINDER_ASSIGNMENT) ? "modulation" : "division";
+        String denominatorName = "this expression";
+        if (denominator.is(Tree.Kind.IDENTIFIER)) {
+          denominatorName = "'" + ((IdentifierTree) denominator).name() + "'";
+        }
+        context.reportIssue(denominator, this, "Make sure " + denominatorName + " can't be zero before doing this " + operation + ".");
+
+        // interrupt exploration
+        return null;
+      }
+    }
+
+    PreStatementVisitor visitor = new PreStatementVisitor(context);
+    syntaxNode.accept(visitor);
+    return visitor.programState;
+  }
+
+  private static boolean isZero(ProgramState programState, SymbolicValue symbolicValue) {
+    return programState.getConstraintWithStatus(symbolicValue, Status.ZERO) != null;
+  }
+
+  private class PreStatementVisitor extends CheckerTreeNodeVisitor {
+
+    PreStatementVisitor(CheckerContext context) {
+      super(context.getState());
+    }
+
+    private void addZeroConstraint(boolean isZero, Tree tree) {
+      initialConstraints.put(tree, isZero ? ZeroConstraint.zero(tree) : ZeroConstraint.nonZero(tree));
```

Comment:
As discussed, this is utterly wrong : something that has not the zero constraint is not necessarily nonZero.